roachtest: add INSERT ... SELECT ... table copy test

nvb opened the pull request:
Closes #15713.

This change adds a test that first imports a fully-populated Bank table. It
then creates a second empty Bank schema. Finally, it performs a series of
paginated `INSERT ... SELECT ...` statements to copy all data from the first
table into the second table. This stresses both large transactions and the
ability for range splits to keep up with a large amount of data movement.

Release note: None

nvb committed:
roachtest: add INSERT ... SELECT ... table copy test

Closes #15713.

This change adds a test that first imports a fully-populated Bank table. It
then creates a second empty Bank schema. Finally, it performs a series of
paginated `INSERT ... SELECT ...` statements to copy all data from the first
table into the second table. This stresses both large transactions and the
ability for range splits to keep up with a large amount of data movement.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23449)
<!-- Reviewable:end -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/roachtest/copy.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-L6tQ1IGmgfPZoogJvR6:-L6tQ1IGmgfPZoogJvR7:b-6604l0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2865800c2bf4feca975353701ce48721d626f5/pkg/cmd/roachtest/copy.go#L43)):*
> ```Go
> 		m := newMonitor(ctx, c, c.All())
> 		m.Go(func(ctx context.Context) error {
> 			run := func(stmt string) {
> ```

Use the `c.Conn` that you're creating below instead.

---

*[pkg/cmd/roachtest/copy.go, line 100 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-L6tQH6yaqRdC4gjgcaZ:-L6tQH6z45E2F0BuDMLQ:byoh4ag) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2865800c2bf4feca975353701ce48721d626f5/pkg/cmd/roachtest/copy.go#L100)):*
> ```Go
> 
> 			// TODO(nvanbenschoten): should we assert that the number of ranges
> 			// in the new table is above some threshold?
> ```

In light of the recent runaway split test, that would be nice.

---

*[pkg/cmd/roachtest/copy.go, line 107 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-L6tQMpi2NyUtdQkSm-4:-L6tQMpi2NyUtdQkSm-5:b-yzqqu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2865800c2bf4feca975353701ce48721d626f5/pkg/cmd/roachtest/copy.go#L107)):*
> ```Go
> 	}
> 
> 	rows := 10000000
> ```

1E7 is a lot easier for me to make sense of.

---

*[pkg/cmd/roachtest/copy.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-L6tQm3CkKRw6SFcO6r0:-L6tQm3CkKRw6SFcO6r1:b2ngyfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2865800c2bf4feca975353701ce48721d626f5/pkg/cmd/roachtest/copy.go#L114)):*
> ```Go
> 		func(t *test) {
> 			if local {
> 				rows = 1000000
> ```

1E6

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-:-L6tQoOYWtzoSx1y_GyI:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/copy.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-L6tQ1IGmgfPZoogJvR6:-L6uDkli4Pf97OwqPIIv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2865800c2bf4feca975353701ce48721d626f5/pkg/cmd/roachtest/copy.go#L43)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Use the `c.Conn` that you're creating below instead.
</blockquote></details>

Done.

---

*[pkg/cmd/roachtest/copy.go, line 100 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-L6tQH6yaqRdC4gjgcaZ:-L6wKOTGrsiat-HmT0JU:b-mqpd2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2865800c2bf4feca975353701ce48721d626f5/pkg/cmd/roachtest/copy.go#L100)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

In light of the recent runaway split test, that would be nice.
</blockquote></details>

Any opinion on the best way to make this assertion? The way I have it now is that it estimates the lower and upper expected bounds for the number of ranges that should exist given the size of the payload and an estimate for the rest of the overhead for a single row. That seems a little fragile though.

---

*[pkg/cmd/roachtest/copy.go, line 107 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-L6tQMpi2NyUtdQkSm-4:-L6uDlEg8PzzGfrJOnIK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2865800c2bf4feca975353701ce48721d626f5/pkg/cmd/roachtest/copy.go#L107)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

1E7 is a lot easier for me to make sense of.
</blockquote></details>

Done.

---

*[pkg/cmd/roachtest/copy.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-L6tQm3CkKRw6SFcO6r0:-L6uDlgB6H5EH3QOJ9tM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2865800c2bf4feca975353701ce48721d626f5/pkg/cmd/roachtest/copy.go#L114)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

1E6
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I'm not seeing the edits you `Done`d, but I assume you'll push them.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/cmd/roachtest/copy.go, line 100 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-L6tQH6yaqRdC4gjgcaZ:-L6wZkC2LXiWk1oiwqIw:b-jydikf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2865800c2bf4feca975353701ce48721d626f5/pkg/cmd/roachtest/copy.go#L100)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Any opinion on the best way to make this assertion? The way I have it now is that it estimates the lower and upper expected bounds for the number of ranges that should exist given the size of the payload and an estimate for the rest of the overhead for a single row. That seems a little fragile though.
</blockquote></details>

I guess ideally you would aggregate the MVCCStats for the table and use that to compute the number of expected ranges. I think it's fine to give what you're doing a try first, though. Your estimate doesn't have to be very exact, we just want to make sure the range count isn't completely off the mark.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-:-L6wZyZJSo-gY15oQmzE:bkc6bsb)*
<!-- Sent from Reviewable.io -->

nvb committed:
roachtest: add progress tracking for operations

This change adds a `Progress` method to `roachtest.test` which records
the progress of the current operation. The progress is reset on the next
call to `Status`.

Release note: None

nvanbenschoten commented:
Just pushed the update. PTAL.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/cmd/roachtest/copy.go, line 100 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-L6tQH6yaqRdC4gjgcaZ:-L6xmhenDMAdjVL3pD5z:b-8esi3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b2865800c2bf4feca975353701ce48721d626f5/pkg/cmd/roachtest/copy.go#L100)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I guess ideally you would aggregate the MVCCStats for the table and use that to compute the number of expected ranges. I think it's fine to give what you're doing a try first, though. Your estimate doesn't have to be very exact, we just want to make sure the range count isn't completely off the mark.
</blockquote></details>

Yeah, you're right that it doesn't need to be exact. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-:-L6xmeKoaRCUcxLljxJU:bbw9fh9)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 1 of 1 files at r2.
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/cmd/roachtest/copy.go, line 103 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-L6yyAO0Ur9Y80T7aR8P:-L6yyAO15Js1hagsqRSr:b-r1wcl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0a9ecfbf9e57d2abb4ede0728448ff18ecb04d/pkg/cmd/roachtest/copy.go#L103)):*
> ```Go
> 			rc := rangeCount()
> 			c.l.printf("range count after copy = %d\n", rc)
> 			highExp := (rows * rowEstimate) / (32 << 20 /* 32MB */)
> ```

That range is approximately [40,80] for these parameters, right? Seems reasonable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23449#-:-L6yybAT7pNhoxdx6Nkv:b-uwkowd)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/cmd/roachtest/copy.go, line 103 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23449#-L6yyAO0Ur9Y80T7aR8P:-L6z1xcF4B8OebqzpCpj:b-km9ibg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0a9ecfbf9e57d2abb4ede0728448ff18ecb04d/pkg/cmd/roachtest/copy.go#L103)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

That range is approximately [40,80] for these parameters, right? Seems reasonable.
</blockquote></details>

Yeah, and the test was consistently ending up with 56 ranges in the new table!

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23449)*
<!-- Sent from Reviewable.io -->

jseldess commented:
@nvanbenschoten, should we resolve or somehow alter this known limitation? https://www.cockroachlabs.com/docs/dev/known-limitations.html#write-and-update-limits-for-a-single-transaction

tschottdorf commented:
@jseldess I don't think we can quite remove it, but it looks wanting an update. I think the current state is

- `kv.raft.command.max_size` is still in effect, though @andreimatei mentioned that we would get better at batching deletions (and perhaps inserts too?) so less and less workloads would actually be able to create large individual Raft commands (note that this is not the same as having a large transactions: many large txns will never create large Raft commands, but the reverse is usually true)
- `kv.transaction.max_intents` has been removed; transactions that write lots of intents work a lot better now thanks to @spencerkimball's changes
- still, long-running write-heavy transactions may never finish and it's good to avoid them, for example via pagination and breaking them up into multiple transactions whenever that is at all possible

@nvanbenschoten this test fails if you don't paginate, right? Ideally we just wouldn't ever send that large a batch down to Raft. It seems that the SQL layer could preclude that without too much trouble?

nvanbenschoten commented:
> transactions that write lots of intents work a lot better now thanks to @spencerkimball's changes

I meant to run these paginated INSERTs both inside and outside of a transaction. As you mentioned, the former should also work now. I'll make that change in a new PR.

> this test fails if you don't paginate, right?

If I don't paginate and I don't increase `kv.raft.command.max_size`, then yes it fails.

> Ideally we just wouldn't ever send that large a batch down to Raft. It seems that the SQL layer could preclude that without too much trouble?

As you mentioned, https://github.com/cockroachdb/cockroach/pull/23373 should get us most of the way here. I'll extend this test further once that change is in to see if it works without pagination.
